modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/SourceConnectionManager.java

```diff
@@ -62,9 +71,21 @@
    * @param connection the connection to be released
    */
   void release(Object connection) {
-    ConnectionHandler<Object> connectionHandler = connections.remove(new Reference<>(connection));
-    if (connectionHandler != null) {
-      connectionHandler.release();
+    Reference<Object> connReference = new Reference<>(connection);
```

Comment:
do we really need to synchronize here? I tend to think no, but yet again, here is where the race condition was.

Reply:
If an AtomicInteger is used, these synchronized blocks can be avoided.